compiler/resolution/lowerIterators.cpp

```diff
@@ -1907,6 +1905,27 @@ replaceErrorFormalWithEnclosingError(SymExpr* se) {
 // out of the enclosing loop. See also the PR message for #12963.
 //
 
+// If we are in a recursive iterator, an IBB may return to the end of the
+// enclosing function. Redirect it to the end of the iterator instead.
+// See #18218.
+//
+static void adjustIbbGotoTarget(GotoStmt* gt, DefExpr*& gtTarget,
+                                Expr* loopRef) {
+  // Normally, the goto's target is in the forLoop's function.
+  if (gtTarget->parentSymbol == loopRef->parentSymbol)
+    return;
+
+  // If this is not for a recursive iterator, let us know.
+  INT_ASSERT(!strncmp(loopRef->parentSymbol->name, "_rec_", 5));
+
+  LabelSymbol* redirect = ((FnSymbol*)(loopRef->parentSymbol))->
```

Comment:
Is there any time this cast could break? Could we do a `toFnSymbol()` instead just to have it crash if that's not the case?